cpp/dolfin/io/XDMFFile.cpp

```diff
@@ -386,8 +391,47 @@ void XDMFFile::write(const function::Function& u)
       = cell_centred ? mesh.num_entities_global(mesh.topology().dim())
                      : num_points;
 
+#ifdef PETSC_USE_COMPLEX
+  std::vector<std::string> components = {"real", "imag"};
+  for (const std::string component : components)
+  {
+    std::string attr_name = component + "_" + u.name();
+    // Add attribute node of the current component
+    pugi::xml_node attribute_node = grid_node.append_child("Attribute");
+    assert(attribute_node);
+    attribute_node.append_attribute("Name") = attr_name.c_str();
+    attribute_node.append_attribute("AttributeType")
+        = rank_to_string(u.value_rank()).c_str();
+    attribute_node.append_attribute("Center") = cell_centred ? "Cell" : "Node";
+
+    // FIXME: Avoid copies by writing directly a compound data
+    std::vector<double> component_data_values(data_values.size());
+    for (unsigned int i = 0; i < data_values.size(); i++)
+    {
+      if (component == components[0])
+        component_data_values[i] = data_values[i].real();
+      else if (component == components[1])
+        component_data_values[i] = data_values[i].imag();
+    }
+
+    // Add data item of component
+    add_data_item(_mpi_comm.comm(), attribute_node, h5_id,
+                  "/VisualisationVector/" + component + "/0",
+                  component_data_values, {num_values, width});
+  }
+
+#else
+  // Add attribute node
+  pugi::xml_node attribute_node = grid_node.append_child("Attribute");
+  assert(attribute_node);
+  attribute_node.append_attribute("Name") = u.name().c_str();
+  attribute_node.append_attribute("AttributeType")
+      = rank_to_string(u.value_rank()).c_str();
+  attribute_node.append_attribute("Center") = cell_centred ? "Cell" : "Node";
```

Comment:
This code block seems the same as above in the component-for-loop. Would it makes sense to factor it out, maybe into a small private function?

Reply:
Or maybe it could work like this:
```
#ifdef PETSC_USE_COMPLEX
  std::vector<std::string> components = {"_real", "_imag"};
#else
  std::vector<std::string> components = {""};
#endif
for (const std::string component : components)
{
  // Common code
}
```

Reply:
Yes, reusing makes code look cleaner.
I selected the second option for all cases that it can apply.